src/Raven.Server/Documents/Handlers/BatchHandler.cs

```diff
@@ -285,16 +266,127 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
                 context.Write(writer, new DynamicJsonValue
                 {
                     [nameof(BatchCommandResult.Results)] = array,
-                    [nameof(BatchCommandResult.TransactionIndex)] = result.Index
+                    [nameof(BatchCommandResult.TransactionIndex)] = index
                 });
             }
         }
 
+        private async Task<long?> GetClusterTransactionCount(object result, string raftRequestId, long databaseCommandsCount, Task<long?> onDatabaseCompletionTask)
+        {
+            if (databaseCommandsCount == 0)
+                return null;
+
+            long? count = null;
+
+            if (result != null && result is ClusterTransactionResult clusterTxResult)
+            {
+                if(clusterTxResult.Count.HasValue == false)
+                    throw new InvalidOperationException($"Cluster Transaction result is null, but has more then 0 database commands.");
+
+                count = clusterTxResult.Count.Value;
+            
+                Database.ForTestingPurposes?.AfterCommitInClusterTransaction?.Invoke();
+            
+                using (var token = CreateHttpRequestBoundOperationToken())
+                    await onDatabaseCompletionTask.WithCancellation(token.Token);
+            
+            }
+            else // leader isn't updated
+            {
+                using (var token = CreateHttpRequestBoundOperationToken())
+                    count = await onDatabaseCompletionTask.WithCancellation(token.Token);
+
+                if (count.HasValue == false) // Failover and the ClusterTransactionCommand is already completed in the DocumentDatabase, so the task with the result is no longer exists, so we'll try to take the result from the local history log.
+                {
+                    using (ServerStore.Engine.ContextPool.AllocateOperationContext(out ClusterOperationContext ctx))
+                    using (ctx.OpenReadTransaction())
+                    {
+                        // searching in the history log
+                        if (ServerStore.Engine.LogHistory.TryGetResultByGuid(ctx, raftRequestId, out var localCommand))
+                        { 
+                            if(localCommand.TryGet(nameof(LogHistoryColumn.Result), out BlittableJsonReaderObject ctlrBlt) == false) // Should never happen! (we are getting here after the command is completed int the db, so it ha been applied and should have results).
+                                throw new InvalidOperationException("Cluster-transaction was succeeded, but Leader is outdated and its results are unaccessible (the command doen't contain 'Result' field).");
```

Comment:
let's add the `localCommand` to this message

Reply:
Done